beacon-chain/blockchain/service.go

```diff
@@ -120,23 +137,16 @@ func (c *ChainService) ProcessedBlockHashes() [][32]byte {
 	return c.processedBlockHashes
 }
 
-// ProcessedCrystallizedStateHashes exposes a getter for the processed crystallized state hashes of the chain.
-func (c *ChainService) ProcessedCrystallizedStateHashes() [][32]byte {
-	return c.processedCrystallizedStateHashes
-}
-
-// ProcessedActiveStateHashes exposes a getter for the processed active state hashes of the chain.
-func (c *ChainService) ProcessedActiveStateHashes() [][32]byte {
-	return c.processedActiveStateHashes
-}
-
 // ProcessBlock accepts a new block for inclusion in the chain.
-func (c *ChainService) ProcessBlock(block *types.Block) error {
+func (c *ChainService) ProcessBlock(block *types.Block) {
 	var canProcess bool
 	var err error
 	h, err := block.Hash()
 	if err != nil {
-		return fmt.Errorf("could not hash incoming block: %v", err)
+		log.Debugf("Could not hash incoming block: %v", err)
+	}
+	if block.SlotNumber() > c.lastSlot && block.SlotNumber() > 1 {
```

Comment:
We need to establish patterns for how to safely access properties concurrently. Take this one for example: `ProcessBlock` is called by the sync service and checks the value of `c.lastSlot`, but `blockProcessing` could easily be modifying it at the same time, leading to a nasty race condition.

Not sure what the actionable item is yet, but ideally these fields are grouped together into distinct goroutines so that we don't have to use locks.

Reply:
This is excellent feedback @rawfalafel, thank you. Completely agree with moving everything into one routine. When you say:

> writing to disk happens in the first loop, and then state transition happens in the second loop

do you mean having a goroutine with different select cases handling each of these items?

Reply:
yea, that's what i meant

Reply:
`updateHead` should come after `c.chain.CanProcessBlock`. Doesn't make sense to call `updateHead` if validation fails.

My understanding of the order of operations:
1. Validate new block
2. Persist block to disk (not state)
3. Run fork choice rule
4. Update active/crystallized state if the head changes

Does that sound correct?

Reply:
Great, addressed this!